README.md
<p align="center">
    <img width="300" height="600" src="https://github.com/tuankhaiit/KTDrawer/raw/master/example/images/layout.gif"/>
</p>

# KTDrawerMenu
[![version](https://img.shields.io/badge/version-0.0.6-green.svg)](https://pub.dev/packages/kt_drawer_menu)

KTDrawerMenu is a library for flutter to create a beautiful drawer menu with deep customization capabilities.


## Installation

Add 
```bash
kt_drawer_menu: ^lastest_version
```
to your pubspec.yaml and run 
```bash
flutter pub get 
```
in your project's root directory.


## Example

```Dart

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(
        brightness: Brightness.light,
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MainPage(),
    );
  }
}

class MainPage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return KTDrawerMenu(
      width: 360.0,
      radius: 30.0,
      scale: 0.6,
      shadow: 20.0,
      shadowColor: Colors.black12,
      drawer: DrawerPage(),
      content: HomePage(),
    );
  }
}

```

[Full Example](https://github.com/tuankhaiit/kt_drawer_menu/blob/master/example/lib/main.dart)


## Available settings

### Controller
This actions is only accessible by the children of KTDrawerMenu.

#### Get controller
```Dart
KTDrawerMenu.of(context);
```
#### Open drawer
```Dart
KTDrawerMenu.of(context).openDrawer();
```
#### Close drawer
```Dart
KTDrawerMenu.of(context).closeDrawer();
```
#### Toggle drawer
```Dart
KTDrawerMenu.of(context).toggle();
```

### Drawer
* Set edge drag width.
* Set width of the drawer widget that which can be displayed.

### Content
* Set scale of content widget when it is opened.
* Set radius of content widget when it is opened.
* Set color opacity above content widget when it is opened.
* Set opacity of color when it is opened.
* Set offset to determine the action auto open/close drawer when the tap is stopped.

### Animation
* Change duration for the open/close animation.
* Listen the animation progress updated when drawer is opening/closing.
* Listen drawer state.


CHANGELOG.md
## [0.0.6] - 10/09/2020.

* Optimized performance.


## [0.0.5+2] - 10/09/2020.

* Change default color to transparent.


## [0.0.5] - 09/09/2020.

* Change default color to transparent.
* Formatted according to dartfmt.


## [0.0.4] - 09/09/2020.

* Update README.md


## [0.0.3] - 09/09/2020.

* Update README.md


## [0.0.2] - 09/09/2020.

* Update README.md


## [0.0.1] - 09/09/2020.

* Start submit first version.


test/kt_drawer_menu_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:kt_drawer_menu/kt_drawer_menu.dart';

void main() {

}


lib/kt_drawer_menu.dart
library kt_drawer_menu;

import 'package:flutter/material.dart';

const kDuration = 180;
const kEdgeDragWidth = 20.0;
const kWidth = 300.0;
const kScale = 0.6;
const kOffset = 0.5;
const kRadius = 40.0;
const kShadow = 10.0;
const kShadowColor = Colors.black38;
const kOpacity = 0.0;
const kColorOverlay = Colors.transparent;

// ignore: must_be_immutable
class KTDrawerMenu extends StatefulWidget {
  final KTDrawerController controller;
  final Widget drawer;
  final Widget content;
  final Duration duration;

  final double edgeDragWidth;
  final double width;
  final double scale;
  final double offset;
  final double radius;
  final double shadow;
  final Color shadowColor;
  final double opacity;
  final Color colorOverlay;
  final AnimationStatusListener onStateChange;
  final ValueChanged<double> onProgressChange;

  KTDrawerMenu({
    Key key,
    @required this.drawer,
    @required this.content,
    KTDrawerController controller,
    Duration duration,
    this.width = kWidth,
    this.edgeDragWidth = kEdgeDragWidth,
    this.scale = kScale,
    this.offset = kOffset,
    this.radius = kRadius,
    this.shadow = kShadow,
    this.shadowColor = kShadowColor,
    this.opacity = kOpacity,
    this.colorOverlay = kColorOverlay,
    this.onStateChange,
    this.onProgressChange,
  })  : this.controller = controller ?? KTDrawerController(),
        this.duration = duration ?? Duration(milliseconds: kDuration),
        assert(edgeDragWidth > 0),
        assert(opacity >= 0 && opacity < 1.0),
        assert(scale > 0 && scale < 1.0),
        assert(offset > 0 && offset < 1.0),
        super(key: key);

  @override
  State<StatefulWidget> createState() => _KTDrawerMenuState();

  static KTDrawerController of(BuildContext context) {
    return context
        .dependOnInheritedWidgetOfExactType<_KTDrawerInheritedWidget>()
        .data;
  }
}

class _KTDrawerMenuState extends State<KTDrawerMenu>
    with TickerProviderStateMixin {
  AnimationController _animationController;
  Animation<double> _animation, _scaleAnimation;
  Animation<double> _alphaAnimation, _shadowAnimation;
  Animation<BorderRadius> _radiusAnimation;

  static final kVelocity = 700.0;
  var _downOffset = Offset.zero;
  var _isDragging = false;
  var isOpen = false;

  @override
  void initState() {
    widget.controller._attach(this);
    super.initState();

    _animationController = AnimationController(
      duration: widget.duration,
      vsync: this,
    );

    _animation =
        Tween<double>(begin: 0.0, end: 1.0).animate(_animationController)
          ..addListener(() {
            if (widget.onProgressChange != null) {
              widget.onProgressChange(_animation.value);
            }
            setState(() {});
          })
          ..addStatusListener((status) {
            if (widget.onStateChange != null) {
              widget.onStateChange(status);
            }
          });

    _scaleAnimation = Tween<double>(begin: 1.0, end: widget.scale)
        .animate(_animationController);

    _alphaAnimation = Tween<double>(begin: 0.0, end: widget.opacity)
        .animate(_animationController);

    _radiusAnimation = BorderRadiusTween(
      begin: BorderRadius.circular(0.0),
      end: BorderRadius.circular(widget.radius),
    ).animate(
        CurvedAnimation(parent: _animationController, curve: Curves.ease));

    _shadowAnimation = Tween<double>(begin: 0.0, end: widget.shadow).animate(
        CurvedAnimation(parent: _animationController, curve: Curves.ease));
  }

  @override
  void didUpdateWidget(KTDrawerMenu oldWidget) {
    widget.controller._attach(this);
    super.didUpdateWidget(oldWidget);
  }

  @override
  void dispose() {
    _animationController.dispose();
    widget.controller._detach(this);
    super.dispose();
  }

  toggle() {
    _animation.isCompleted ? closeDrawer() : openDrawer();
  }

  closeDrawer() {
    _animationController.reverse();
    isOpen = false;
  }

  openDrawer() {
    _animationController.forward();
    isOpen = true;
  }

  @override
  Widget build(BuildContext context) {
    var device = MediaQuery.of(context).size;
    var width = device.width;
    var height = device.height;

    return _KTDrawerInheritedWidget(
      data: widget.controller,
      child: Material(
        type: MaterialType.transparency,
        child: GestureDetector(
          onHorizontalDragDown: (DragDownDetails details) {
            var position = details.globalPosition;
            if (isOpen || (!isOpen && position.dx < widget.edgeDragWidth)) {
              _downOffset = position;
              _isDragging = true;
            }
          },
          onHorizontalDragUpdate: (DragUpdateDetails details) {
            var position = details.globalPosition;
            if (_isDragging) {
              if (isOpen) {
                var newValue =
                    1 - (_downOffset.dx - position.dx - 15) / widget.width;
                _animationController.value = newValue;
              } else {
                var newValue =
                    (position.dx - _downOffset.dx - 15) / widget.width;
                _animationController.value = newValue;
              }
            }
          },
          onHorizontalDragEnd: (DragEndDetails details) {
            if (_isDragging) {
              if (details != null && details.primaryVelocity != null) {
                if (details.primaryVelocity > kVelocity) {
                  openDrawer();
                } else if (details.primaryVelocity < -kVelocity) {
                  closeDrawer();
                } else {
                  if (_animationController.value > widget.offset) {
                    openDrawer();
                  } else {
                    closeDrawer();
                  }
                }
              }
            }

            _isDragging = false;
          },
          child: Stack(
            children: [
              Container(
                color: Colors.white,
                width: width,
                height: height,
                child: widget.drawer,
              ),
              Transform.scale(
                scale: _scaleAnimation.value,
                child: Transform.translate(
                  offset: Offset(widget.width * _animation.value, 0.0),
                  child: Container(
                    width: width,
                    height: height,
                    decoration: BoxDecoration(
                        borderRadius: _radiusAnimation.value,
                        boxShadow: [
                          BoxShadow(
                              color: widget.shadow == 0
                                  ? Colors.transparent
                                  : widget.shadowColor.withOpacity(0.5),
                              spreadRadius: _shadowAnimation.value / 2.0,
                              blurRadius: _shadowAnimation.value * 2)
                        ]),
                    child: ClipRRect(
                      borderRadius: _radiusAnimation.value,
                      child: Stack(
                        children: [
                          widget.content,
                          Container(
                            color: _animation.value == 0
                                ? null
                                : widget.colorOverlay
                                    .withOpacity(_alphaAnimation.value),
                          )
                        ],
                      ),
                    ),
                  ),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

class _KTDrawerInheritedWidget extends InheritedWidget {
  final KTDrawerController data;

  _KTDrawerInheritedWidget({
    Key key,
    @required Widget child,
    @required this.data,
  }) : super(key: key, child: child);

  @override
  bool updateShouldNotify(InheritedWidget oldWidget) {
    return true;
  }
}

typedef KTDrawerStateCallback = void Function(bool isOpen);
typedef KTDrawerUpdateCallback = void Function(double progress);

class KTDrawerController {
  _KTDrawerMenuState _state;

  KTDrawerController();

  bool isOpening() => _state.isOpen;

  void openDrawer() {
    if (_state != null) {
      _state.openDrawer();
    }
  }

  void closeDrawer() {
    if (_state != null) {
      _state.closeDrawer();
    }
  }

  void toggle() {
    if (_state != null) {
      _state.toggle();
    }
  }

  void _attach(_KTDrawerMenuState state) {
    _state = state;
  }

  void _detach(_KTDrawerMenuState state) {
    _state = null;
  }
}


example/lib/main.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:kt_drawer_menu/kt_drawer_menu.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        brightness: Brightness.light,
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MainPage(),
    );
  }
}

class MainPage extends StatelessWidget {
  // ignore: close_sinks
  final StreamController<DrawerItemEnum> _streamController =
      StreamController<DrawerItemEnum>.broadcast(sync: true);

  @override
  Widget build(BuildContext context) {
    return KTDrawerMenu(
      width: 360.0,
      radius: 30.0,
      scale: 0.6,
      shadow: 20.0,
      shadowColor: Colors.black12,
      drawer: DrawerPage(streamController: _streamController),
      content: HomePage(streamController: _streamController),
    );
  }
}

enum DrawerItemEnum {
  DASHBOARD,
  MESSAGE,
  SETTINGS,
  ABOUT,
  HELP,
}

// ignore: must_be_immutable
class DrawerPage extends StatelessWidget {
  final StreamController<DrawerItemEnum> streamController;
  DrawerItemEnum selected;

  DrawerPage({Key key, this.streamController}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return StreamBuilder<DrawerItemEnum>(
      stream: streamController.stream,
      initialData: DrawerItemEnum.DASHBOARD,
      builder: (context, snapshot) {
        selected = snapshot.data;
        return Container(
          color: Colors.blueGrey[900],
          child: Stack(
            children: [
              Center(
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    _getMenu(context, DrawerItemEnum.DASHBOARD),
                    _getMenu(context, DrawerItemEnum.MESSAGE),
                    _getMenu(context, DrawerItemEnum.SETTINGS),
                    _getMenu(context, DrawerItemEnum.ABOUT),
                    _getMenu(context, DrawerItemEnum.HELP),
                  ],
                ),
              )
            ],
          ),
        );
      },
    );
  }

  Widget _getMenu(BuildContext context, DrawerItemEnum menu) {
    switch (menu) {
      case DrawerItemEnum.DASHBOARD:
        return _buildItem(context, menu, "Dashboard", Icons.dashboard, () {});
      case DrawerItemEnum.MESSAGE:
        return _buildItem(context, menu, "Messages", Icons.message, () {});
      case DrawerItemEnum.SETTINGS:
        return _buildItem(context, menu, "Settings", Icons.settings, () {});
      case DrawerItemEnum.ABOUT:
        return _buildItem(context, menu, "About", Icons.info, () {});
      case DrawerItemEnum.HELP:
        return _buildItem(context, menu, "Help", Icons.help_outline, () {});
      default:
        return Container();
    }
  }

  Widget _buildItem(
    BuildContext context,
    DrawerItemEnum menu,
    String title,
    IconData icon,
    Function onPressed,
  ) {
    return Material(
      color: Colors.transparent,
      child: InkWell(
        splashColor: Colors.transparent,
        onTap: () {
          streamController.sink.add(menu);
          KTDrawerMenu.of(context).closeDrawer();
          onPressed();
        },
        child: Container(
          height: 50,
          padding: EdgeInsets.only(left: 26),
          child: Row(
            children: [
              Icon(icon,
                  color: selected == menu ? Colors.white : Colors.white70,
                  size: 24),
              SizedBox(width: 14),
              Text(
                title,
                style: TextStyle(
                  color: Colors.white,
                  fontSize: selected == menu ? 15 : 14,
                  fontWeight:
                      selected == menu ? FontWeight.w500 : FontWeight.w300,
                ),
              )
            ],
          ),
        ),
      ),
    );
  }
}

class HomePage extends StatefulWidget {
  final StreamController<DrawerItemEnum> streamController;

  HomePage({Key key, this.streamController}) : super(key: key);

  @override
  State<StatefulWidget> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  DrawerItemEnum selected;

  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return StreamBuilder(
      stream: widget.streamController.stream,
      initialData: DrawerItemEnum.DASHBOARD,
      builder: (context, snapshot) {
        selected = snapshot.data;
        return Scaffold(
          appBar: AppBar(
            backgroundColor: _getColor(selected),
            elevation: 0.0,
            leading: IconButton(
              icon: Icon(Icons.dehaze),
              onPressed: () {
                KTDrawerMenu.of(context).toggle();
              },
            ),
            title: Text(
              _getTitle(selected),
              style: TextStyle(
                color: Colors.white,
              ),
            ),
          ),
          body: Container(
            color: _getColor(selected),
          ),
        );
      },
    );
  }

  Color _getColor(DrawerItemEnum menu) {
    switch (menu) {
      case DrawerItemEnum.DASHBOARD:
        return Colors.blue;
      case DrawerItemEnum.MESSAGE:
        return Colors.green;
      case DrawerItemEnum.SETTINGS:
        return Colors.purple;
      case DrawerItemEnum.ABOUT:
        return Colors.orange;
      case DrawerItemEnum.HELP:
        return Colors.yellow;
      default:
        return Colors.white;
    }
  }

  String _getTitle(DrawerItemEnum menu) {
    switch (menu) {
      case DrawerItemEnum.DASHBOARD:
        return "DASHBOARD";
      case DrawerItemEnum.MESSAGE:
        return "MESSAGE";
      case DrawerItemEnum.SETTINGS:
        return "SETTINGS";
      case DrawerItemEnum.ABOUT:
        return "ABOUT";
      case DrawerItemEnum.HELP:
        return "HELP";
      default:
        return "";
    }
  }
}
